misk/src/test/kotlin/misk/hibernate/TransacterTest.kt

```diff
@@ -13,23 +14,83 @@ class TransacterTest {
+    // Insert some movies, characters and actors.
     transacter.transaction { session ->
-      session.save(DbMovie("Jurassic Park", Date()))
-      session.save(DbMovie("Star Wars", Date()))
+      val jp = session.save(DbMovie("Jurassic Park", LocalDate.of(1993, 6, 9), clock.instant()))
+      val sw = session.save(DbMovie("Star Wars", LocalDate.of(1977, 5, 25), clock.instant()))
+      assertThat(setOf(jp, sw)).hasSize(2) // Uniqueness check.
+
+      val ld = session.save(DbActor("Laura Dern", LocalDate.of(1967, 2, 10)))
+      val jg = session.save(DbActor("Jeff Goldblum", LocalDate.of(1952, 10, 22)))
+      val cf = session.save(DbActor("Carrie Fisher", null))
+      assertThat(setOf(ld, jg, cf)).hasSize(3) // Uniqueness check.
+
+      val ah = session.save(DbCharacter("Amilyn Holdo", session.load(sw), session.load(ld)))
+      val es = session.save(DbCharacter("Ellie Sattler", session.load(jp), session.load(ld)))
+      val im = session.save(DbCharacter("Ian Malcolm", session.load(jp), session.load(jg)))
+      val lo = session.save(DbCharacter("Leia Organa", session.load(sw), session.load(cf)))
+      assertThat(setOf(ah, es, im, lo)).hasSize(4) // Uniqueness check.
     }
 
-    // Query some movies.
+    // Query that data.
     transacter.transaction { session ->
-      val criteriaBuilder = session.newCriteriaBuilder()
-      val criteria = criteriaBuilder.createQuery(DbMovie::class.java)
-      val queryRoot = criteria.from(DbMovie::class.java)
-      criteria.where(criteriaBuilder.notEqual(queryRoot.get<String>("name"), "Star Wars"))
-      val resultList: List<DbMovie> = session.query(criteria).list()
-      assertThat(resultList.map { it.name }).containsExactly("Jurassic Park")
+      val ianMalcolm = queryFactory.newQuery<CharacterQuery>()
+          .name("Ian Malcolm")
+          .uniqueResult(session)!!
+      assertThat(ianMalcolm.actor.name).isEqualTo("Jeff Goldblum")
+      assertThat(ianMalcolm.movie.name).isEqualTo("Jurassic Park")
+
+      val lauraDernMovies = queryFactory.newQuery<CharacterQuery>()
+          .actorName("Laura Dern")
+          .listAs<MovieNameAndReleaseDate>(session)
+      assertThat(lauraDernMovies).containsExactly(
+          MovieNameAndReleaseDate("Star Wars", LocalDate.of(1977, 5, 25)),
+          MovieNameAndReleaseDate("Jurassic Park", LocalDate.of(1993, 6, 9)))
+
+      val actorsInOldMovies = queryFactory.newQuery<CharacterQuery>()
+          .movieReleaseDateBefore(LocalDate.of(1980, 1, 1))
+          .listAs<ActorAndReleaseDate>(session)
+      assertThat(actorsInOldMovies).containsExactly(
+          ActorAndReleaseDate("Laura Dern", LocalDate.of(1977, 5, 25)),
+          ActorAndReleaseDate("Carrie Fisher", LocalDate.of(1977, 5, 25)))
     }
   }
+
+  annotation class Constraint(
+    val path: String,
+    val operator: String = "="
+  )
+
+  annotation class Property(
+    val value: String
+  )
+
+  interface CharacterQuery : Query<DbCharacter> {
```

Comment:
should this live outside of the test?

Reply:
tough. The point of the test is to confirm that the query bindings work.